File: mobile/test/api_client_test.dart
import 'dart:convert';

import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:nexusvpn/core/api_client.dart';
import 'package:nexusvpn/core/app_exception.dart';

/// Seeds the secure store so ApiClient sees a signed-in session.
void seedSession({String access = 'access-1', String refresh = 'refresh-1'}) {
  FlutterSecureStorage.setMockInitialValues({
    'nexusvpn.access_token': access,
    'nexusvpn.refresh_token': refresh,
    'nexusvpn.server_url': 'https://api.example.com/api/v1',
  });
}

http.Response jsonResponse(Object body, {int status = 200}) =>
    http.Response(jsonEncode(body), status, headers: {'content-type': 'application/json'});

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() => FlutterSecureStorage.setMockInitialValues({}));

  test('login posts credentials and returns the token pair', () async {
    seedSession();
    late http.Request captured;

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((req) async {
        captured = req;
        return jsonResponse({
          'accessToken': 'a-1',
          'refreshToken': 'r-1',
          'expiresIn': 900,
        });
      }),
    );

    final tokens = await client.login(email: 'alice@example.com', password: 'correct-horse-battery');

    expect(tokens.accessToken, 'a-1');
    expect(captured.method, 'POST');
    expect(captured.url.path, '/api/v1/auth/login');
    expect(jsonDecode(captured.body)['email'], 'alice@example.com');
  });

  test('login surfaces the MFA challenge rather than treating it as success', () async {
    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((_) async => jsonResponse({'mfaRequired': true})),
    );

    final tokens = await client.login(email: 'alice@example.com', password: 'pw');

    expect(tokens.mfaRequired, isTrue);
    expect(tokens.hasTokens, isFalse);
  });

  test('attaches the bearer token to authenticated requests', () async {
    seedSession(access: 'the-access-token');
    String? authHeader;

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((req) async {
        authHeader = req.headers['authorization'];
        return jsonResponse([]);
      }),
    );

    await client.listNetworks();

    expect(authHeader, 'Bearer the-access-token');
  });

  test('refreshes once on 401 and retries the original request', () async {
    seedSession(access: 'stale-token');
    var networkCalls = 0;
    var refreshCalls = 0;

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((req) async {
        if (req.url.path.endsWith('/auth/refresh')) {
          refreshCalls++;
          return jsonResponse({
            'accessToken': 'fresh-token',
            'refreshToken': 'r-2',
            'expiresIn': 900,
          });
        }
        networkCalls++;
        if (req.headers['authorization'] == 'Bearer fresh-token') {
          return jsonResponse([
            {'id': 'n-1', 'name': 'Home Lab', 'cidr': '10.77.0.0/24'},
          ]);
        }
        return http.Response('unauthorized', 401);
      }),
    );

    final networks = await client.listNetworks();

    expect(networks, hasLength(1));
    expect(networks.first.name, 'Home Lab');
    expect(refreshCalls, 1, reason: 'exactly one refresh should be attempted');
    expect(networkCalls, 2, reason: 'the original request should be retried once');
  });

  test('signals session expiry when the refresh token is also rejected', () async {
    seedSession(access: 'stale-token');
    var sessionExpired = false;

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((_) async => http.Response('unauthorized', 401)),
    )..onSessionExpired = () => sessionExpired = true;

    await expectLater(client.listNetworks(), throwsA(isA<ApiException>()));
    expect(sessionExpired, isTrue);
  });

  test('surfaces the API error envelope', () async {
    seedSession();

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((_) async => jsonResponse(
            {
              'error': {'code': 'not_found', 'message': 'Network not found'}
            },
            status: 404,
          )),
    );

    try {
      await client.getNetwork('missing');
      fail('expected an ApiException');
    } on ApiException catch (e) {
      expect(e.statusCode, 404);
      expect(e.message, contains('Network not found'));
    }
  });

  test('joins a network by invite code', () async {
    seedSession();
    late http.Request captured;

    final client = ApiClient(
      baseUrl: 'https://api.example.com/api/v1',
      httpClient: MockClient((req) async {
        captured = req;
        return jsonResponse({'id': 'n-9', 'name': 'Team', 'cidr': '10.80.0.0/24'});
      }),
    );

    final network = await client.joinNetwork('INVITE-CODE');

    expect(network.id, 'n-9');
    expect(captured.url.path, '/api/v1/networks/join');
    expect(jsonDecode(captured.body)['inviteCode'], 'INVITE-CODE');
  });
}

File: client/internal/apiclient/apiclient_test.go
package apiclient

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
)

func TestLogin_StoresTokens(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/auth/login" || r.Method != http.MethodPost {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body map[string]string
		json.NewDecoder(r.Body).Decode(&body)
		if body["email"] != "user@example.com" {
			t.Fatalf("unexpected email: %s", body["email"])
		}
		json.NewEncoder(w).Encode(TokenPair{AccessToken: "access-1", RefreshToken: "refresh-1", ExpiresIn: 900})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	tokens, err := c.Login(context.Background(), "user@example.com", "hunter2", "")
	if err != nil {
		t.Fatalf("Login: %v", err)
	}
	if tokens.AccessToken != "access-1" {
		t.Fatalf("AccessToken = %q, want access-1", tokens.AccessToken)
	}
	tok, _ := c.AccessToken(context.Background())
	if tok != "access-1" {
		t.Fatalf("stored access token = %q, want access-1", tok)
	}
}

func TestLogin_MFARequiredDoesNotStoreTokens(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		json.NewEncoder(w).Encode(TokenPair{MFARequired: true})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	tokens, err := c.Login(context.Background(), "user@example.com", "hunter2", "")
	if err != nil {
		t.Fatalf("Login: %v", err)
	}
	if !tokens.MFARequired {
		t.Fatalf("expected MFARequired true")
	}
	tok, _ := c.AccessToken(context.Background())
	if tok != "" {
		t.Fatalf("expected no access token to be stored when MFA required, got %q", tok)
	}
}

func TestLogin_InvalidCredentials(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]string{"code": "invalid_credentials", "message": "bad email or password"},
		})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	_, err := c.Login(context.Background(), "user@example.com", "wrong", "")
	if err == nil {
		t.Fatalf("expected error")
	}
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != http.StatusUnauthorized || apiErr.Code != "invalid_credentials" {
		t.Fatalf("unexpected APIError: %+v", apiErr)
	}
}

func TestAuthenticatedRequest_AttachesBearerToken(t *testing.T) {
	var gotAuth atomic.Value
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth.Store(r.Header.Get("Authorization"))
		json.NewEncoder(w).Encode([]Network{{ID: "net-1", Name: "Home"}})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	c.SetTokens("my-access-token", "my-refresh-token")

	networks, err := c.ListNetworks(context.Background())
	if err != nil {
		t.Fatalf("ListNetworks: %v", err)
	}
	if len(networks) != 1 || networks[0].ID != "net-1" {
		t.Fatalf("unexpected networks: %+v", networks)
	}
	if gotAuth.Load().(string) != "Bearer my-access-token" {
		t.Fatalf("Authorization header = %q", gotAuth.Load())
	}
}

func TestAuthenticatedRequest_RefreshesOn401(t *testing.T) {
	var calls int32
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/networks":
			n := atomic.AddInt32(&calls, 1)
			if n == 1 {
				// First call with the stale token: unauthorized.
				if r.Header.Get("Authorization") != "Bearer stale-token" {
					t.Fatalf("expected stale token on first call, got %q", r.Header.Get("Authorization"))
				}
				w.WriteHeader(http.StatusUnauthorized)
				json.NewEncoder(w).Encode(map[string]any{"error": map[string]string{"code": "unauthorized"}})
				return
			}
			// Retried call after refresh should carry the new token.
			if r.Header.Get("Authorization") != "Bearer fresh-token" {
				t.Fatalf("expected fresh token on retry, got %q", r.Header.Get("Authorization"))
			}
			json.NewEncoder(w).Encode([]Network{{ID: "net-1"}})
		case "/auth/refresh":
			json.NewEncoder(w).Encode(TokenPair{AccessToken: "fresh-token", RefreshToken: "fresh-refresh"})
		default:
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
	}))
	defer srv.Close()

	var refreshedTokens TokenPair
	c := New(srv.URL, Options{OnTokenRefresh: func(tp TokenPair) { refreshedTokens = tp }})
	c.SetTokens("stale-token", "some-refresh-token")

	networks, err := c.ListNetworks(context.Background())
	if err != nil {
		t.Fatalf("ListNetworks: %v", err)
	}
	if len(networks) != 1 {
		t.Fatalf("unexpected networks: %+v", networks)
	}
	if atomic.LoadInt32(&calls) != 2 {
		t.Fatalf("expected 2 calls to /networks (original + retry), got %d", calls)
	}
	if refreshedTokens.AccessToken != "fresh-token" {
		t.Fatalf("OnTokenRefresh callback not invoked with new tokens: %+v", refreshedTokens)
	}
}

func TestJoinNetwork(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/networks/join" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		var body map[string]string
		json.NewDecoder(r.Body).Decode(&body)
		if body["inviteCode"] != "ABC123" {
			t.Fatalf("unexpected invite code: %s", body["inviteCode"])
		}
		json.NewEncoder(w).Encode(Network{ID: "net-1", Name: "Home Lab", CIDR: "10.77.0.0/24"})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	c.SetTokens("tok", "ref")

	network, err := c.JoinNetwork(context.Background(), "ABC123")
	if err != nil {
		t.Fatalf("JoinNetwork: %v", err)
	}
	if network.ID != "net-1" || network.CIDR != "10.77.0.0/24" {
		t.Fatalf("unexpected network: %+v", network)
	}
}

func TestRegisterDevice(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(Device{ID: "dev-1", VirtualIP: "10.77.0.9", PublicKey: "pubkey"})
	}))
	defer srv.Close()

	c := New(srv.URL, Options{})
	c.SetTokens("tok", "ref")

	device, err := c.RegisterDevice(context.Background(), "net-1", "laptop", "linux", "6.1", "pubkey")
	if err != nil {
		t.Fatalf("RegisterDevice: %v", err)
	}
	if device.VirtualIP != "10.77.0.9" {
		t.Fatalf("unexpected device: %+v", device)
	}
}
